File: connect/mcp/src/ohd_connect_mcp/openfoodfacts.py
"""OpenFoodFacts (OFF) resolver — used by the MCP food-resolution tools.

Mirrors the proxy-first / direct-fallback behaviour of the Android
``OpenFoodFacts.kt`` resolver (``connect/android/.../data/OpenFoodFacts.kt``)
so an LLM agent can chain ``food_lookup_barcode`` / ``food_search`` into
``log_food`` when the user names a product without giving full nutrition.

Order of resolution:

1. OHD SaaS proxy at ``https://api.ohd.dev/v1/openfoodfacts`` (short timeout;
   not deployed today, fails fast).
2. Public OFF v2 / search-CGI endpoints directly.

Returns dicts (not pydantic models) so the MCP tool surface stays JSON-native;
the LLM is the consumer and reads field names directly.
"""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

# OFF asks API consumers to identify themselves in User-Agent.
USER_AGENT = "OHD-Connect-MCP/0.1 (+https://ohd.dev)"

PROXY_PRODUCT = "https://api.ohd.dev/v1/openfoodfacts"
PROXY_SEARCH = "https://api.ohd.dev/v1/openfoodfacts/search"

OFF_PRODUCT = "https://world.openfoodfacts.org/api/v2/product"
OFF_SEARCH = "https://world.openfoodfacts.org/cgi/search.pl"

# Intentionally no ``fields=`` selector. Early prototypes used one with
# ``fields=...nutriments,nutriments_100g,...`` — the bogus ``nutriments_100g``
# name made OFF substitute ``nutrition_data*`` markers for the whole nested
# ``nutriments`` object, zeroing every macro on every product. Pulling the
# full payload also matches the user's intent (E-numbers, caffeine, vitamins,
# NOVA / Nutri-Score, ingredient list — everything goes into food.eaten).

# Proxy timeouts are short — falling through fast is the right behaviour
# while api.ohd.dev hasn't shipped. OFF direct gets more headroom.
PROXY_CONNECT_S = 2.0
PROXY_READ_S = 3.0
OFF_CONNECT_S = 4.0
OFF_READ_S = 8.0


async def lookup_barcode(barcode: str) -> dict[str, Any] | None:
    """Resolve an EAN/UPC to a product dict, or ``None`` if neither endpoint knows it."""
    cleaned = (barcode or "").strip()
    if not cleaned.isdigit() or not (6 <= len(cleaned) <= 14):
        return None

    proxy_url = f"{PROXY_PRODUCT}/{cleaned}"
    body = await _http_get(proxy_url, PROXY_CONNECT_S, PROXY_READ_S)
    if body is not None:
        product = _parse_product(body, source="OpenFoodFacts (via OHD proxy)")
        if product is not None:
            return product

    off_url = f"{OFF_PRODUCT}/{cleaned}.json"
    body = await _http_get(off_url, OFF_CONNECT_S, OFF_READ_S)
    if body is None:
        return None
    return _parse_product(body, source="OpenFoodFacts")


async def search(query: str, page_size: int = 10) -> list[dict[str, Any]]:
    """Free-text product search. Empty list when both endpoints fail or no hits."""
    cleaned = (query or "").strip()
    if not cleaned:
        return []
    page_size = max(1, min(page_size, 50))

    encoded = quote(cleaned, safe="")
    proxy_url = f"{PROXY_SEARCH}?query={encoded}&page_size={page_size}"
    body = await _http_get(proxy_url, PROXY_CONNECT_S, PROXY_READ_S)
    if body is not None:
        hits = _parse_search(body, source="OpenFoodFacts (via OHD proxy)")
        if hits:
            return hits

    off_url = (
        f"{OFF_SEARCH}?search_terms={encoded}&search_simple=1"
        f"&action=process&json=1&page_size={page_size}"
    )
    body = await _http_get(off_url, OFF_CONNECT_S, OFF_READ_S)
    if body is None:
        return []
    return _parse_search(body, source="OpenFoodFacts")


async def _http_get(url: str, connect_s: float, read_s: float) -> dict[str, Any] | None:
    timeout = httpx.Timeout(connect=connect_s, read=read_s, write=read_s, pool=read_s)
    headers = {"User-Agent": USER_AGENT, "Accept": "application/json"}
    try:
        async with httpx.AsyncClient(timeout=timeout, follow_redirects=True) as client:
            response = await client.get(url, headers=headers)
    except (httpx.HTTPError, OSError):
        return None
    if response.status_code < 200 or response.status_code >= 300:
        return None
    try:
        return response.json()
    except ValueError:
        return None


def _parse_product(body: dict[str, Any], source: str) -> dict[str, Any] | None:
    """OFF v2 product GET → flat dict. Returns ``None`` if status≠1 or no product."""
    status = body.get("status")
    if status != 1:
        return None
    product = body.get("product")
    if not isinstance(product, dict):
        return None
    return _map_product(product, source=source)


def _parse_search(body: dict[str, Any], source: str) -> list[dict[str, Any]]:
    products = body.get("products")
    if not isinstance(products, list):
        return []
    out: list[dict[str, Any]] = []
    for entry in products:
        if not isinstance(entry, dict):
            continue
        name = (entry.get("product_name") or "").strip()
        brand = (entry.get("brands") or "").strip()
        if not name and not brand:
            continue
        out.append(_map_product(entry, source=source))
    return out


def _map_product(product: dict[str, Any], source: str) -> dict[str, Any]:
    name_raw = (product.get("product_name") or "").strip()
    brands_raw = (product.get("brands") or "").strip()
    code = (product.get("code") or "").strip()

    if name_raw:
        name = name_raw
    elif brands_raw:
        name = brands_raw
    else:
        name = f"Unknown product (barcode {code})" if code else "Unknown product"

    brand = brands_raw.split(",", 1)[0].strip() if brands_raw else None

    nutriments = product.get("nutriments")
    if not isinstance(nutriments, dict):
        nutriments = {}

    # Sodium: prefer `sodium_100g` (grams → mg), else approximate from salt (÷2.5).
    sodium_g = _f(nutriments, "sodium_100g")
    salt_g = _f(nutriments, "salt_100g")
    if sodium_g > 0:
        sodium_mg: float = sodium_g * 1_000.0
    elif salt_g > 0:
        sodium_mg = (salt_g / 2.5) * 1_000.0
    else:
        sodium_mg = 0.0

    per_100g = {
        "kcal": int(_f(nutriments, "energy-kcal_100g")),
        "carbs_g": _f(nutriments, "carbohydrates_100g"),
        "protein_g": _f(nutriments, "proteins_100g"),
        "fat_g": _f(nutriments, "fat_100g"),
        "sugar_g": _f(nutriments, "sugars_100g"),
        "fiber_g": _f(nutriments, "fiber_100g"),
        "saturated_fat_g": _f(nutriments, "saturated-fat_100g"),
        "trans_fat_g": _f(nutriments, "trans-fat_100g"),
        "sodium_mg": sodium_mg,
        "cholesterol_mg": _f(nutriments, "cholesterol_100g") * 1_000.0,
        "potassium_mg": _f(nutriments, "potassium_100g") * 1_000.0,
        "calcium_mg": _f(nutriments, "calcium_100g") * 1_000.0,
        "iron_mg": _f(nutriments, "iron_100g") * 1_000.0,
        "vitamin_c_mg": _f(nutriments, "vitamin-c_100g") * 1_000.0,
        "vitamin_d_mcg": _f(nutriments, "vitamin-d_100g") * 1_000_000.0,
        "caffeine_mg": _f(nutriments, "caffeine_100g") * 1_000.0,
    }

    nova = product.get("nova_group")
    nova_int: int | None
    if isinstance(nova, int) and 1 <= nova <= 4:
        nova_int = nova
    elif isinstance(nova, str):
        try:
            parsed = int(nova)
            nova_int = parsed if 1 <= parsed <= 4 else None
        except ValueError:
            nova_int = None
    else:
        nova_int = None

    nutri = (product.get("nutriscore_grade") or "").strip().lower() or None
    if nutri in {"unknown", "not-applicable"}:
        nutri = None
    eco = (product.get("ecoscore_grade") or "").strip().lower() or None
    if eco in {"unknown", "not-applicable"}:
        eco = None

    return {
        "barcode": code or None,
        "name": name,
        "brand": brand,
        "source": source,
        "description": _describe(product),
        "package_serving": _package_serving(product),
        "per_100g": per_100g,
        "image_thumb_url": (product.get("image_thumb_url") or None),
        # Composition — every E-number, every allergen, every analysis tag.
        # Empty lists when OFF didn't have data; never None so JSON consumers
        # can rely on the shape.
        "additives": _strip_tags(product.get("additives_tags")),
        "allergens": _strip_tags(product.get("allergens_tags")),
        "traces": _strip_tags(product.get("traces_tags")),
        "ingredients": _strip_tags(product.get("ingredients_tags"), limit=50),
        "ingredients_analysis": _strip_tags(product.get("ingredients_analysis_tags")),
        "labels": _strip_tags(product.get("labels_tags"), limit=20),
        "nova_group": nova_int,
        "nutri_score": nutri,
        "eco_score": eco,
        # Every ``*_100g`` nutriment OFF returned, in OFF's native units. The
        # LLM picks what to surface — caffeine, omega-3, phosphorus, etc.
        "all_nutriments_per_100g": _harvest_all(nutriments),
    }


def _harvest_all(nutriments: dict[str, Any]) -> dict[str, float]:
    """Snapshot every ``*_100g`` nutriment, stripping the suffix."""
    out: dict[str, float] = {}
    for key in nutriments:
        if not isinstance(key, str) or not key.endswith("_100g"):
            continue
        if key == "energy_100g" or key.startswith("nutrition-score"):
            continue
        value = _f(nutriments, key)
        if value == 0.0:
            continue
        out[key.removesuffix("_100g")] = value
    return out


def _strip_tags(tags: Any, limit: int | None = None) -> list[str]:
    """OFF tag arrays → plain lowercase slugs, dropping the ``en:`` prefix."""
    if not isinstance(tags, list):
        return []
    out: list[str] = []
    for tag in tags:
        if not isinstance(tag, str) or not tag:
            continue
        idx = tag.find(":")
        out.append(tag[idx + 1:] if 0 < idx <= 3 else tag)
        if limit is not None and len(out) >= limit:
            break
    return out


def _describe(product: dict[str, Any]) -> str:
    generic = (product.get("generic_name") or "").strip()
    if generic:
        return generic
    tags = product.get("categories_tags")
    if not isinstance(tags, list):
        return ""
    cleaned: list[str] = []
    for tag in tags:
        if not isinstance(tag, str) or not tag:
            continue
        # Drop language prefix like "en:" → keep slug.
        idx = tag.find(":")
        cleaned.append(tag[idx + 1:] if 0 < idx <= 3 else tag)
    return ", ".join(cleaned[-3:])


def _package_serving(product: dict[str, Any]) -> dict[str, Any] | None:
    label = (product.get("quantity") or "").strip()
    raw = product.get("product_quantity")
    grams: float | None
    if raw is None:
        grams = None
    elif isinstance(raw, (int, float)):
        grams = float(raw)
    elif isinstance(raw, str):
        stripped = raw.strip()
        try:
            grams = float(stripped) if stripped else None
        except ValueError:
            grams = None
    else:
        grams = None
    if not label and grams is None:
        return None
    if not label and grams is not None:
        label = f"{int(grams)} g"
    return {"name": label, "grams": grams or 0.0}


def _f(obj: dict[str, Any], key: str) -> float:
    """Like ``JSONObject.optDouble`` — missing/non-numeric → 0.0."""
    v = obj.get(key)
    if isinstance(v, (int, float)):
        return float(v)
    if isinstance(v, str):
        try:
            return float(v)
        except ValueError:
            return 0.0
    return 0.0

File: connect/android/app/src/main/java/com/ohd/connect/data/OpenFoodFacts.kt
package com.ohd.connect.data

import com.ohd.connect.ui.screens.FoodItem
import com.ohd.connect.ui.screens.NutritionFacts
import com.ohd.connect.ui.screens.Serving
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runInterruptible
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ConcurrentHashMap

/**
 * OpenFoodFacts (OFF) barcode resolver — spec §4.7.
 *
 * Triggered from [com.ohd.connect.ui.screens.FoodSearchScreen] when the query
 * looks like an EAN/UPC (8–13 digits) and no local match was found in
 * [com.ohd.connect.ui.screens.FoodDictionary].
 *
 * Tries two endpoints in order:
 *   1. OHD SaaS proxy `https://api.ohd.dev/v1/openfoodfacts/{barcode}` —
 *      not deployed yet, so we keep the connect timeout short (~2 s) and fall
 *      through on any network / 4xx / 5xx failure.
 *   2. OpenFoodFacts v2 directly:
 *      `https://world.openfoodfacts.org/api/v2/product/{barcode}.json` with a
 *      pruned `?fields=...` selection to keep payloads small.
 *
 * Uses only `java.net.HttpURLConnection` + `org.json.JSONObject` to avoid
 * adding new HTTP dependencies. All network I/O runs on `Dispatchers.IO`.
 *
 * The OFF terms-of-service ask for a short rate limit between calls; the
 * search screen only fires a single lookup per query keystroke, so we don't
 * implement explicit throttling here.
 */
object OpenFoodFacts {

    /** User-Agent header — required by OFF's API ToS. */
    private const val USER_AGENT = "OHD-Connect/0.1 (Android; +https://ohd.dev)"

    /** OHD proxy endpoint. Not yet deployed — must fail-fast. */
    private const val PROXY_BASE = "https://api.ohd.dev/v1/openfoodfacts"

    /** OHD proxy text-search endpoint. Same caveat as [PROXY_BASE]. */
    private const val PROXY_BASE_SEARCH = "https://api.ohd.dev/v1/openfoodfacts/search"

    /** OpenFoodFacts v2 product endpoint. */
    private const val OFF_BASE = "https://world.openfoodfacts.org/api/v2/product"

    /** OpenFoodFacts text-search endpoint (legacy cgi but still stable). */
    private const val OFF_SEARCH_BASE = "https://world.openfoodfacts.org/cgi/search.pl"

    // Intentionally no `fields=` selector — early betas used one and it
    // silently zeroed nutrition for every scan: with `fields=...,nutriments,
    // nutriments_100g,...` OFF substitutes `nutrition_data*` markers and
    // strips the entire nested `nutriments` object. Pulling the full product
    // is also what the user actually wants — every micronutrient, every
    // E-number, every label travels into the food.eaten payload.

    // Connect timeout stays short so an unreachable proxy fails over to the
    // direct OFF call fast. Read timeout has headroom — the proxy IS live now
    // (api.ohd.dev) and a full search response is 10 large product objects.
    private const val PROXY_CONNECT_TIMEOUT_MS = 2_000
    private const val PROXY_READ_TIMEOUT_MS = 8_000

    // Direct OFF call gets more headroom — public endpoint, occasionally slow.
    private const val OFF_CONNECT_TIMEOUT_MS = 4_000
    private const val OFF_READ_TIMEOUT_MS = 8_000

    /**
     * Process-lifetime cache so [com.ohd.connect.ui.screens.FoodDetailScreen]
     * can re-resolve a remote item by name after navigation. Keyed by
     * barcode; values are the mapped [FoodItem].
     */
    val cache: ConcurrentHashMap<String, FoodItem> = ConcurrentHashMap()

    /**
     * Resolve a barcode to a [FoodItem]. Returns `null` if neither the proxy
     * nor OFF directly recognises the code.
     *
     * Never throws — network / parse exceptions are swallowed so the caller
     * can fall back to the "not found" UI. (Errors that *should* be visible
     * are surfaced separately by [lookupWithSource]; this top-level wrapper
     * is kept exception-free for parity with the spec's `lookup` signature.)
     */
    suspend fun lookup(barcode: String): FoodItem? = runInterruptible(Dispatchers.IO) {
        // Wrapped in `runInterruptible` so coroutine cancellation (user types
        // another character, navigates away, etc.) translates to a thread
        // interrupt, which `HttpURLConnection` honours by throwing
        // `InterruptedIOException` → swallowed by `runCatching`. Without this
        // the blocking I/O outlived the composition and Compose logged a
        // "coroutine left the composition" warning.
        //
        // The api.ohd.dev proxy is currently misconfigured (serves OFF's HTML
        // error page for every path) — go straight to OFF. Re-add the proxy
        // first-hop once the Caddy route is fixed + caching is wired.
        runCatching { fetchOff(barcode) }.getOrNull()
    }

    /**
     * Free-text product search against OpenFoodFacts (or the OHD proxy if it
     * ever comes online — the proxy URL slot is reserved but not yet
     * deployed). Falls through to the public OFF search endpoint when the
     * proxy is unavailable.
     *
     * Returns up to [pageSize] hits, ranked by OFF's default popularity
     * heuristic. Every result is also written to [cache] so a subsequent
     * tap → [com.ohd.connect.ui.screens.FoodDetailScreen] resolves locally.
     */
    suspend fun search(
        query: String,
        pageSize: Int = 10,
    ): List<FoodItem> = runInterruptible(Dispatchers.IO) {
        val cleaned = query.trim()
        if (cleaned.isBlank()) return@runInterruptible emptyList()
        // Proxy skipped — see [lookup] note. Direct OFF text search.
        runCatching { searchOff(cleaned, pageSize) }.getOrNull().orEmpty()
    }

    /**
     * Try the proxy first, returning `(item, sourceLabel)`. Returns `null`
     * only if both endpoints fail. Used by tests / debug screens; the UI uses
     * [lookup] which discards the source label (it's already on `FoodItem.source`).
     */
    suspend fun lookupWithSource(barcode: String): FoodItem? = lookup(barcode)

    // ------------------------------------------------------------------
    // Endpoint wrappers
    // ------------------------------------------------------------------

    private fun fetchProxy(barcode: String): FoodItem? {
        val url = URL("$PROXY_BASE/$barcode")
        val body = httpGet(url, PROXY_CONNECT_TIMEOUT_MS, PROXY_READ_TIMEOUT_MS)
            ?: return null
        val item = parseOffJson(body, "OpenFoodFacts (via OHD proxy)") ?: return null
        cache[barcode] = item
        return item
    }

    private fun fetchOff(barcode: String): FoodItem? {
        val url = URL("$OFF_BASE/$barcode.json")
        val body = httpGet(url, OFF_CONNECT_TIMEOUT_MS, OFF_READ_TIMEOUT_MS)
            ?: return null
        val item = parseOffJson(body, "OpenFoodFacts") ?: return null
        cache[barcode] = item
        return item
    }

    // ------------------------------------------------------------------
    // Search endpoints
    // ------------------------------------------------------------------

    private fun searchProxy(query: String, pageSize: Int): List<FoodItem> {
        // The proxy is a dumb path-rewrite (/v1/openfoodfacts/search →
        // /cgi/search.pl) — it forwards the query string verbatim. So we
        // send OFF's *native* search params, not a `query=` alias, otherwise
        // cgi/search.pl serves an HTML landing page instead of JSON.
        val encoded = java.net.URLEncoder.encode(query, "UTF-8")
        val url = URL(
            "$PROXY_BASE_SEARCH?search_terms=$encoded&search_simple=1" +
                "&action=process&json=1&page_size=$pageSize",
        )
        val body = httpGet(url, PROXY_CONNECT_TIMEOUT_MS, PROXY_READ_TIMEOUT_MS)
            ?: return emptyList()
        return parseOffSearchJson(body, "OpenFoodFacts (via OHD proxy)")
    }

    private fun searchOff(query: String, pageSize: Int): List<FoodItem> {
        // OFF's `cgi/search.pl` endpoint takes a search-term + JSON flag and
        // accepts the same `fields=` selector as the product GET. It's the
        // stable text-search surface; the newer `search.openfoodfacts.org`
        // is for the SaaS demo site and not guaranteed.
        val encoded = java.net.URLEncoder.encode(query, "UTF-8")
        val url = URL(
            "$OFF_SEARCH_BASE?search_terms=$encoded&search_simple=1" +
                "&action=process&json=1&page_size=$pageSize",
        )
        val body = httpGet(url, OFF_CONNECT_TIMEOUT_MS, OFF_READ_TIMEOUT_MS)
            ?: return emptyList()
        return parseOffSearchJson(body, "OpenFoodFacts")
    }

    /**
     * Parse OFF's search.pl response: `{ "products": [ {...}, ... ] }`.
     * Each entry is the same product shape as the v2 product endpoint, so
     * we reuse [mapOffProduct]. Hits with a missing/blank `product_name` AND
     * blank `brands` are dropped to avoid surfacing junk rows.
     */
    private fun parseOffSearchJson(body: String, sourceLabel: String): List<FoodItem> {
        val root = runCatching { JSONObject(body) }.getOrNull() ?: return emptyList()
        val products = root.optJSONArray("products") ?: return emptyList()
        val out = ArrayList<FoodItem>(products.length())
        for (i in 0 until products.length()) {
            val product = products.optJSONObject(i) ?: continue
            val name = product.optString("product_name").trim()
            val brand = product.optString("brands").trim()
            if (name.isEmpty() && brand.isEmpty()) continue
            val item = mapOffProduct(product, sourceLabel)
            cache[item.name] = item   // keyed by name so foodByName(name) resolves
            val code = product.optString("code").trim()
            if (code.isNotEmpty()) cache[code] = item
            out += item
        }
        return out
    }

    // ------------------------------------------------------------------
    // HTTP
    // ------------------------------------------------------------------

    /**
     * Plain GET. Returns the response body on 2xx, `null` on any other
     * status / network failure so the caller can fall through.
     */
    private fun httpGet(url: URL, connectTimeoutMs: Int, readTimeoutMs: Int): String? {
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = connectTimeoutMs
            readTimeout = readTimeoutMs
            instanceFollowRedirects = true
            setRequestProperty("User-Agent", USER_AGENT)
            setRequestProperty("Accept", "application/json")
        }
        return try {
            val code = conn.responseCode
            if (code in 200..299) {
                conn.inputStream.bufferedReader().use { it.readText() }
            } else {
                null
            }
        } catch (_: Throwable) {
            null
        } finally {
            conn.disconnect()
        }
    }

    // ------------------------------------------------------------------
    // Parsing — OFF v2 response → FoodItem
    // ------------------------------------------------------------------

    /**
     * Parse an OFF v2 product response and project it into our [FoodItem].
     * Returns `null` if `status != 1` or no `product` field is present.
     *
     * Visible for testing.
     */
    internal fun parseOffJson(body: String, sourceLabel: String): FoodItem? {
        val root = runCatching { JSONObject(body) }.getOrNull() ?: return null
        val status = root.optInt("status", 0)
        if (status != 1) return null
        val product = root.optJSONObject("product") ?: return null
        return mapOffProduct(product, sourceLabel)
    }

    private fun mapOffProduct(product: JSONObject, sourceLabel: String): FoodItem {
        val rawName = product.optString("product_name").trim()
        val brandsRaw = product.optString("brands").trim()
        val code = product.optString("code").trim()

        val name = when {
            rawName.isNotEmpty() -> rawName
            brandsRaw.isNotEmpty() -> brandsRaw
            else -> "Unknown product (barcode $code)"
        }

        val brand = brandsRaw
            .takeIf { it.isNotEmpty() }
            ?.split(",")
            ?.firstOrNull()
            ?.trim()
            ?.takeIf { it.isNotEmpty() }

        val description = buildDescription(product)

        val nutriments = product.optJSONObject("nutriments")

        // Sodium: prefer `sodium_100g` (grams → mg), otherwise approximate
        // from `salt_100g` (grams of salt ≈ 2.5× grams of sodium) → mg.
        val sodiumG = optDouble(nutriments, "sodium_100g")
        val saltG = optDouble(nutriments, "salt_100g")
        val sodiumMg = when {
            sodiumG > 0.0 -> sodiumG * 1_000.0
            saltG > 0.0 -> (saltG / 2.5) * 1_000.0
            else -> 0.0
        }

        val per100g = NutritionFacts(
            kcal = optDouble(nutriments, "energy-kcal_100g").toInt(),
            carbsG = optDouble(nutriments, "carbohydrates_100g"),
            proteinG = optDouble(nutriments, "proteins_100g"),
            fatG = optDouble(nutriments, "fat_100g"),
            sugarG = optDouble(nutriments, "sugars_100g"),
            fiberG = optDouble(nutriments, "fiber_100g"),
            saturatedFatG = optDouble(nutriments, "saturated-fat_100g"),
            transFatG = optDouble(nutriments, "trans-fat_100g"),
            sodiumMg = sodiumMg,
            // Cholesterol / minerals / vitamins: OFF reports in grams, we
            // store in milligrams (×1000). Vitamin D is reported in grams
            // but our UI shows micrograms (×1_000_000) — the conventional
            // dose unit for cholecalciferol.
            cholesterolMg = optDouble(nutriments, "cholesterol_100g") * 1_000.0,
            potassiumMg = optDouble(nutriments, "potassium_100g") * 1_000.0,
            calciumMg = optDouble(nutriments, "calcium_100g") * 1_000.0,
            ironMg = optDouble(nutriments, "iron_100g") * 1_000.0,
            vitaminCMg = optDouble(nutriments, "vitamin-c_100g") * 1_000.0,
            vitaminDMcg = optDouble(nutriments, "vitamin-d_100g") * 1_000_000.0,
            // Caffeine: OFF stores grams/100g; we expose mg/100g.
            caffeineMg = optDouble(nutriments, "caffeine_100g") * 1_000.0,
            nutrimentsPer100g = harvestAllPer100g(nutriments),
        )

        val packageServing = buildPackageServing(product)

        return FoodItem(
            name = name,
            brand = brand,
            source = sourceLabel,
            description = description,
            per100g = per100g,
            packageServing = packageServing,
            defaultPortion = null,
            additives = stringList(product, "additives_tags"),
            allergens = stringList(product, "allergens_tags"),
            traces = stringList(product, "traces_tags"),
            // Cap ingredients at 50 — OFF returns deep hierarchies that bloat
            // the food.eaten event payload (parent categories + leaves +
            // synonyms). 50 is comfortably more than any real ingredient list.
            ingredients = stringList(product, "ingredients_tags", limit = 50),
            ingredientsAnalysis = stringList(product, "ingredients_analysis_tags"),
            labels = stringList(product, "labels_tags", limit = 20),
            novaGroup = product.opt("nova_group").let {
                when (it) {
                    is Number -> it.toInt().takeIf { n -> n in 1..4 }
                    is String -> it.toIntOrNull()?.takeIf { n -> n in 1..4 }
                    else -> null
                }
            },
            nutriScore = product.optString("nutriscore_grade").trim().lowercase()
                .takeIf { it.isNotEmpty() && it != "unknown" && it != "not-applicable" },
            ecoScore = product.optString("ecoscore_grade").trim().lowercase()
                .takeIf { it.isNotEmpty() && it != "unknown" && it != "not-applicable" },
        )
    }

    /**
     * Snapshot every `*_100g` key OFF returned, stripping the suffix. Values
     * stay in OFF's native unit (mostly grams) so downstream code can compute
     * "per gram-eaten" with one rule-of-three.
     *
     * Skips `energy_100g` (kJ duplicate of `energy-kcal_100g` × 4.184) and
     * `nutrition-score-*_100g` (Nutri-Score component, not a nutriment).
     */
    private fun harvestAllPer100g(nutriments: JSONObject?): Map<String, Double> {
        if (nutriments == null) return emptyMap()
        val out = LinkedHashMap<String, Double>()
        val keys = nutriments.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            if (!key.endsWith("_100g")) continue
            if (key == "energy_100g") continue
            if (key.startsWith("nutrition-score")) continue
            val v = optDouble(nutriments, key)
            if (v == 0.0) continue
            val short = key.removeSuffix("_100g")
            out[short] = v
        }
        return out
    }

    /**
     * Pull an OFF tag array (e.g. `additives_tags`) into a plain `List<String>`
     * with the `en:` / `fr:` language prefix stripped. Optionally caps result
     * length to keep payloads manageable on deeply-tagged products.
     */
    private fun stringList(
        product: JSONObject,
        key: String,
        limit: Int = Int.MAX_VALUE,
    ): List<String> {
        val arr = product.optJSONArray(key) ?: return emptyList()
        val out = ArrayList<String>(minOf(arr.length(), 32))
        for (i in 0 until arr.length()) {
            val raw = arr.optString(i).trim()
            if (raw.isEmpty()) continue
            out += stripLangPrefix(raw)
            if (out.size >= limit) break
        }
        return out
    }

    /**
     * `generic_name` → fallback to last 3 `categories_tags` (with the `en:`
     * language prefix stripped) → fallback to empty string.
     */
    private fun buildDescription(product: JSONObject): String {
        val generic = product.optString("generic_name").trim()
        if (generic.isNotEmpty()) return generic

        val tagsArr = product.optJSONArray("categories_tags") ?: return ""
        val tags = (0 until tagsArr.length())
            .mapNotNull { tagsArr.optString(it).takeIf { s -> s.isNotEmpty() } }
            .map { stripLangPrefix(it) }
        if (tags.isEmpty()) return ""
        return tags.takeLast(3).joinToString(", ")
    }

    /** "en:soft-drinks" → "soft-drinks"; "soft-drinks" stays. */
    private fun stripLangPrefix(tag: String): String {
        val idx = tag.indexOf(':')
        return if (idx in 1..3) tag.substring(idx + 1) else tag
    }

    /**
     * Build the optional package-serving from `product.quantity` (label) and
     * `product.product_quantity` (grams). `product_quantity` is sometimes a
     * string, sometimes a number — handle both.
     */
    private fun buildPackageServing(product: JSONObject): Serving? {
        val label = product.optString("quantity").trim()
        val grams: Double? = when (val v = product.opt("product_quantity")) {
            null -> null
            JSONObject.NULL -> null
            is Number -> v.toDouble()
            is String -> v.trim().takeIf { it.isNotEmpty() }?.toDoubleOrNull()
            else -> null
        }
        if (label.isEmpty() && grams == null) return null
        val displayName = label.ifEmpty { grams?.let { "${it.toInt()} g" } ?: return null }
        return Serving(name = displayName, grams = grams ?: 0.0).takeIf { it.grams > 0.0 || label.isNotEmpty() }
    }

    /**
     * `JSONObject.optDouble` returns `NaN` for missing keys; we want 0.0
     * instead (per spec). Also tolerates numeric values stored as strings.
     */
    private fun optDouble(obj: JSONObject?, key: String): Double {
        if (obj == null) return 0.0
        val v = obj.opt(key) ?: return 0.0
        return when (v) {
            is Number -> v.toDouble()
            is String -> v.toDoubleOrNull() ?: 0.0
            else -> 0.0
        }
    }
}
